src/lib/formats/fs_cbmdos.cpp

```diff
@@ -0,0 +1,554 @@
+{
+	std::vector<meta_description> results;
+	results.emplace_back(meta_name::name, "", false, [](const meta_value &m) { return m.as_string().size() <= 16; }, "File name, up to 16 characters");
+	results.emplace_back(meta_name::file_type, "", true, nullptr, "Type of the file");
+	results.emplace_back(meta_name::length, 0, true, nullptr, "Size of the file in bytes");
+	return results;
+}
+
+
+//-------------------------------------------------
+//  mount
+//-------------------------------------------------
+
+std::unique_ptr<filesystem_t> fs::cbmdos_image::mount(fsblk_t &blockdev) const
+{
+	return std::make_unique<impl>(blockdev);
+}
+
+
+//-------------------------------------------------
+//  validate_filename
+//-------------------------------------------------
+
+namespace {
+
+bool validate_filename(std::string_view name)
+{
+	return true;
+}
+
+
+//-------------------------------------------------
+//  strtrimright_cbm
+//-------------------------------------------------
+
+std::string_view strtrimright_cbm(std::string_view str)
+{
+	return strtrimright(str, [](char c) { return c != (char)0xA0; });
+}
+
+
+//-------------------------------------------------
+//  strtrimright_cbm
+//-------------------------------------------------
+
+template<size_t N>
+std::string_view strtrimright_cbm(char str[N])
+{
+	std::string_view sv(str, std::size(str));
+	return strtrimright_cbm(sv);
+}
+
+
+//-------------------------------------------------
+//  impl ctor
+//-------------------------------------------------
+
+impl::impl(fsblk_t &blockdev)
+	: filesystem_t(blockdev, 256)
+{
+}
+
+
+//-------------------------------------------------
+//  impl::metadata
+//-------------------------------------------------
+
+meta_data impl::metadata()
+{
+	auto bam_block = read_sector(18, 0);
+	std::string_view disk_name = std::string_view((const char *) bam_block.rodata() + 0x90, 16);
+
+	meta_data results;
+	results.set(meta_name::name, strtrimright_cbm(disk_name));
+	return results;
+}
+
+
+//-------------------------------------------------
+//  impl::root
+//-------------------------------------------------
+
+filesystem_t::dir_t impl::root()
+{
+	if (!m_root)
+		m_root = new root_dir(*this);
+	return m_root.strong();
+}
+
+
+//-------------------------------------------------
+//  impl::drop_root_ref
+//-------------------------------------------------
+
+void impl::drop_root_ref()
+{
+	m_root = nullptr;
+}
+
+
+//-------------------------------------------------
+//  impl::read_sector
+//-------------------------------------------------
+
+fsblk_t::block_t impl::read_sector(int track, int sector) const
+{
+	// CBM thinks in terms of tracks/sectors, but we have a block device abstraction
+	u32 block = 0;
+	block += std::max(std::min(track, 18) -  1, 0) * 21;
+	block += std::max(std::min(track, 25) - 18, 0) * 19;
+	block += std::max(std::min(track, 31) - 25, 0) * 18;
+	block += std::max(std::min(track, 41) - 31, 0) * 17;
+	block += sector;
```

Comment:
This is an area of MAME that I'm not very familiar with, so I'm not able to provide a useful review, sorry.  However, I wanted to note that the above sector calculations are Commodore 1541 (`c1541` or `D64` format) specific.  I think you're likely aware of that already, and you've only hooked this up to `c1541`.  

Just in case, here's the info on the other formats:
http://www.baltissen.org/newhtm/diskimag.htm

I've used that info on other projects and found it to be accurate.  The formats are used in MAME by at least these drivers:

- `D64`: `c1541` (as you have hooked up), plus `c2031`, `c2040` and `c4040`
- `D71`: `c1571`
- `D81`: `c1581`
- `D80`: `c8050`
- `D82`: `c8250`
- `D90`: `d9060`

Somewhat related: 32d5d9079776ae40c99c840b6acd889cf150a18c surfaced that writing into `D41` and `D71` is broken.  Hopefully that can be fixed at some point.  This PR also doesn't support writing.  I think being able to write files into the images or the filesystem would be very useful.

Reply:
Thanks for the info - I've got my hands full with a number of other FS issues (I want to see FS functionality exposed to LUA scripts), so if someone wants to take these other CBM variations on that would be great.

Of course, this PR has to be merged first